xds/src/main/java/io/grpc/xds/XdsNameResolver2.java

```diff
@@ -0,0 +1,387 @@
+      Metadata headers = args.getHeaders();
+      for (String headerName : headers.keys()) {
+        if (headerName.endsWith(Metadata.BINARY_HEADER_SUFFIX)) {
+          continue;
+        }
+        Metadata.Key<String> key = Metadata.Key.of(headerName, Metadata.ASCII_STRING_MARSHALLER);
+        asciiHeaders.put(headerName, headers.getAll(key));
+      }
+      String cluster = null;
+      Route selectedRoute = null;
+      do {
+        for (Route route : routes) {
+          if (route.getRouteMatch().matches(
+              "/" + args.getMethodDescriptor().getFullMethodName(), asciiHeaders)) {
+            selectedRoute = route;
+            break;
+          }
+        }
+        if (selectedRoute == null) {
+          return Result.forError(Status.UNAVAILABLE.withDescription("Failed to route the RPC"));
+        }
+        RouteAction action = selectedRoute.getRouteAction();
+        if (action.getCluster() != null) {
+          cluster = action.getCluster();
+        } else if (action.getWeightedCluster() != null) {
+          int totalWeight = 0;
+          for (ClusterWeight weightedCluster : action.getWeightedCluster()) {
+            totalWeight += weightedCluster.getWeight();
+          }
+          int select = random.nextInt(totalWeight);
+          int accumulator = 0;
+          for (ClusterWeight weightedCluster : action.getWeightedCluster()) {
+            accumulator += weightedCluster.getWeight();
+            if (select < accumulator) {
+              cluster = weightedCluster.getName();
+              break;
+            }
+          }
+        }
+        if (cluster == null) {  // should not happen if routing rules are configured correctly
+          return Result.forError(
+              Status.UNAVAILABLE.withDescription("Failed to route the RPC with selected action"));
+        }
+      } while (!retainCluster(cluster));
+      final String finalCluster = cluster;
+      class SelectionCompleted implements Runnable {
+        @Override
+        public void run() {
+          releaseCluster(finalCluster);
+        }
+      }
+
+      String serviceConfigJson =
+          generateServiceConfigWithMethodConfig(
+              args.getMethodDescriptor().getFullMethodName(),
+              selectedRoute.getRouteAction().getTimeoutNano());
+      ConfigOrError parsedServiceConfig = parseServiceConfig(serviceConfigJson);
+      Object config = parsedServiceConfig.getConfig();
+      if (config == null) {
+        throw new AssertionError(
+            "Bug: invalid config", parsedServiceConfig.getError().asException());
+      }
+      return
+          Result.newBuilder()
+              .setCallOptions(args.getCallOptions().withOption(CLUSTER_SELECTION_KEY, cluster))
+              .setConfig(config)
+              .setCommittedCallback(new SelectionCompleted())
+              .build();
+    }
+
+    private boolean retainCluster(String cluster) {
+      AtomicInteger refCount = clusterRefs.get(cluster);
```

Comment:
Could this refCount be `null`? A serious problem is that `routes` and `clusterRefs` are not updated atomically. `retainCluster()` is not run in synchronization context, so it's hard for me to verify the correctness.

Reply:
Yeah, there could be race problems. The cluster to retain may be being removed by another call. Fixed here.

Another place is when a new update is received and try to create reference count for new clusters, checking if a cluster reference already exists and creating the reference for it could be interleaved by another thread that deletes the reference for this cluster. Fixed there too.